### ISSUE TITLE
GeoDataFrame Intersection throws ValueError

### ISSUE THREAD

**Issue by erikburt** [geopandas/geopandas#1130]
Issue
-----
When doing a `.overlay(how='intersection')` on a GeoDataFrame, there are inconsistencies resulting in errors. Here is the stack trace:

```
Traceback (most recent call last):
  File "intersection_crash.py", line 17, in <module>
    intersection_big_gdf = gpd.overlay(concave_polygon_gdf, big_rect_polygon_gdf, how='intersection')
  File "/usr/local/lib/python3.6/site-packages/geopandas-0.6.0rc1-py3.6.egg/geopandas/tools/overlay.py", line 362, in overlay
    result = _overlay_intersection(df1, df2)
  File "/usr/local/lib/python3.6/site-packages/geopandas-0.6.0rc1-py3.6.egg/geopandas/tools/overlay.py", line 217, in _overlay_intersection
    intersections = left.intersection(right).buffer(0)
  File "/usr/local/lib/python3.6/site-packages/geopandas-0.6.0rc1-py3.6.egg/geopandas/base.py", line 530, in buffer
    resolution=resolution, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/geopandas-0.6.0rc1-py3.6.egg/geopandas/base.py", line 92, in _delegate_geo_method
    data = getattr(a_this, op)(*args, **kwargs).data
  File "/usr/local/lib/python3.6/site-packages/geopandas-0.6.0rc1-py3.6.egg/geopandas/array.py", line 616, in buffer
    return GeometryArray(np.array(data, dtype=object))
  File "/usr/local/lib/python3.6/site-packages/geopandas-0.6.0rc1-py3.6.egg/geopandas/array.py", line 407, in __init__
    "'data' should be a 1-dimensional array of geometry objects.")
ValueError: 'data' should be a 1-dimensional array of geometry objects.
```

Steps to Reproduce
---------------------
Run the following code we are using python 3.6 and geopandas 0.6.0rc1

```
import pandas as pd
import geopandas as gpd
from shapely.geometry import Polygon

concave_polygon = {'geometry': [Polygon([(-5.509260350101057, 36.260244201586794), (-5.731064912610456, 36.266167159517529), (-5.832776397076094, 36.250849164869088) , (-5.829100078360468, 36.285161472881597), (-5.579110405697939, 36.29190139052691), (-5.182524246397088, 36.260244201586794), (-5.182524246397088, 35.741474782826323), (-5.509260350101057, 35.741474782826323), (-5.509260350101057, 36.260244201586794)])]}
concave_polygon_gdf = gpd.GeoDataFrame(concave_polygon, geometry='geometry')

small_rect_polygon = {'geometry': [Polygon([(-5.509260350101057, 36.260244201586794), (-5.182524246397088, 36.260244201586794), (-5.182524246397088, 35.741474782826323), (-5.509260350101057, 35.741474782826323), (-5.509260350101057, 36.260244201586794)])]}
small_rect_polygon_gdf = gpd.GeoDataFrame(small_rect_polygon, geometry='geometry')

big_rect_polygon = {'geometry': [Polygon([(-6, 36.260244201586794), (-5.182524246397088, 36.260244201586794), (-5.182524246397088, 35.741474782826323), (-6, 35.741474782826323), (-6, 36.260244201586794)])]}
big_rect_polygon_gdf = gpd.GeoDataFrame(big_rect_polygon, geometry='geometry')

intersection_small_gdf = gpd.overlay(concave_polygon_gdf, small_rect_polygon_gdf, how='intersection')

# Error occurs here
intersection_big_gdf = gpd.overlay(concave_polygon_gdf, big_rect_polygon_gdf, how='intersection')
```

Investigation
--------------

The issue appears to be happening because of wrapping the resulting data into a list when it is already one. I'm guessing this is happening around [here](https://github.com/geopandas/geopandas/blob/master/geopandas/array.py#L629).

**Comment by jorisvandenbossche** [comment]
@erikburt Thanks a lot for this detailed and nicely reproducible bug report!

So the problem is in this line in the `buffer` method:

```
return GeometryArray(np.array(data, dtype=object)
```

where `data` is a list. This is correct as such, but the problem is a corner case of numpy, where if that list is a one-element list of an object that is iterable, it will unpack it and create a multi-dimensional array:

```
--> 639         return GeometryArray(np.array(data, dtype=object))
    640 
    641     def interpolate(self, distance, normalized=False):

ipdb> data                                                                                                                                                                                                         
[<shapely.geometry.multipolygon.MultiPolygon object at 0x7f3444122ac8>]
ipdb> np.array(data, dtype=object)                                                                                                                                                                                 
array([[<shapely.geometry.polygon.Polygon object at 0x7f34441225c0>,
        <shapely.geometry.polygon.Polygon object at 0x7f345f7aaa90>]],
      dtype=object)
```

So a solution to solve this is doing the following (first creating empty 1D array and then filling it):

```
arr = np.empty(len(data), dtype=object)
arr[:] = data
```

**Comment by jorisvandenbossche** [comment]
If you want to do a PR to fix this, very welcome to try! (otherwise I will do it in a few days, as this is related to the refactoring, so would like to fix this for including in 0.6.0)
